### PR TITLE
[Endpoint Checks][Autodiscovery] Support check configs defined in files

### PR DESCRIPTION
**Pull request by ahmed-mez**
### What does this PR do?

Support file-based endpoint checks.

### Motivation

Same as https://github.com/DataDog/datadog-agent/pull/10073

### Additional Notes

This is the part 2 of https://github.com/DataDog/datadog-agent/pull/10073

### Describe how to test/QA your changes

- Define endpoint checks with `advanced_ad_identifiers`, create and delete the targeted endpoints multiple times - make sure AD works as expected.
Example: (http check)
```
advanced_ad_identifiers:
  - kube_endpoints:
      name: nginx
      namespace: default
instances:
  - url: "http://%%host%%"
    name: nginx
```
- Endpoint checks based on dd or prometheus annotations should still work as expected
- A mix of service and endpoint checks can also be used and should generate a service cluster check + endpoint checks (as many as the number of endpoints)
```
advanced_ad_identifiers:
  - kube_service:
      name: nginx
      namespace: default
  - kube_endpoints:
      name: nginx
      namespace: default
instances:
  - url: "http://%%host%%"
    name: nginx
```

### Reviewer's Checklist

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [x] The appropriate `team/..` label has been applied, if known.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] Changed code has automated tests for its functionality.
- [x] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [x] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [x] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [x] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Review by alai97 — APPROVED** [review]
Looks good!

**Review by djmitche — APPROVED** [review]
👍🏽  for pkg/config/autodiscovery/autodiscovery.go

File: pkg/autodiscovery/providers/kube_endpoints_file.go [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+	return configs
+}
+
+func epID(ns, name string) string {
+	return ns + "/" + name
+}
+
+func init() {
+	RegisterProvider(names.KubeEndpointsFileRegisterName, NewKubeEndpointsFileConfigProvider)
+}
```
Comment (clamoriniere) [inline pkg/autodiscovery/providers/kube_endpoints_file.go:310]: could you move the `init()` at the top of the file?

File: pkg/config/autodiscovery/autodiscovery.go [hunk 2]
```diff
@@ -30,13 +30,33 @@ func DiscoverComponentsFromConfig() ([]config.ConfigurationProviders, []config.L
...
 
-	advancedConfigs, _, err := providers.ReadConfigFiles(providers.WithAdvancedADOnly)
-	if err != nil {
-		log.Debugf("Couldn't read config files: %w", err)
-	}
+	// Auto-add file-based kube service and endpoints config providers based on check config files.
+	if flavor.GetFlavor() == flavor.ClusterAgent {
+		advancedConfigs, _, err := providers.ReadConfigFiles(providers.WithAdvancedADOnly)
+		if err != nil {
+			log.Warnf("Couldn't read config files: %w", err)
```
Comment (clamoriniere) [inline pkg/config/autodiscovery/autodiscovery.go:37]: should we continue if `err !=nil` ?

Comment (clamoriniere) [reply pkg/config/autodiscovery/autodiscovery.go:37]: It means that `advancedConfigs` is also nil or empty

File: pkg/config/autodiscovery/autodiscovery.go [hunk 3]
```diff
@@ -30,13 +30,33 @@ func DiscoverComponentsFromConfig() ([]config.ConfigurationProviders, []config.L
...
+					log.Info("Configs with advanced kube service identifiers detected: Adding the 'kube service file' config provider")
+					detectedProviders = append(detectedProviders, config.ConfigurationProviders{Name: names.KubeServicesFileRegisterName, Polling: false})
+				}
 
-	if len(advancedConfigs) > 0 && flavor.GetFlavor() == flavor.ClusterAgent {
-		detectedProviders = append(detectedProviders, config.ConfigurationProviders{Name: "kube_services_file", Polling: false})
+				if !epFound && !adv.KubeEndpoints.IsEmpty() {
+					epFound = true
+					log.Info("Configs with advanced kube endpoints identifiers detected: Adding the 'kube endpoints file' config provider")
+					detectedProviders = append(detectedProviders, config.ConfigurationProviders{Name: names.KubeEndpointsFileRegisterName, Polling: true})
```
Comment (clamoriniere) [inline pkg/config/autodiscovery/autodiscovery.go:52]: could you add a comment that explain why `Polling: true` for `endpoints` but not for `service`

File: pkg/autodiscovery/providers/kube_endpoints_file.go (hunk 1 above)
Comment (ahmed-mez) [reply pkg/autodiscovery/providers/kube_endpoints_file.go]: done

File: pkg/config/autodiscovery/autodiscovery.go (hunk 3 above)
Comment (ahmed-mez) [reply pkg/config/autodiscovery/autodiscovery.go:57]: done

File: pkg/autodiscovery/listeners/kube_endpoints.go [hunk 4]
```diff
@@ -64,7 +66,7 @@ func init() {
 	Register(kubeEndpointsName, NewKubeEndpointsListener)
 }
 
-func NewKubeEndpointsListener(Config) (ServiceListener, error) {
+func NewKubeEndpointsListener(conf Config) (ServiceListener, error) {
```
Comment (clamoriniere) [inline pkg/autodiscovery/listeners/kube_endpoints.go:69]: is `Config` an Interface?
if not, should we do instead

```suggestion
func NewKubeEndpointsListener(conf *Config) (ServiceListener, error) {
```

File: pkg/autodiscovery/providers/kube_endpoints_file.go [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+// KubeEndpointsFileConfigProvider generates endpoints checks from check configurations defined in files.
+type KubeEndpointsFileConfigProvider struct {
+	sync.RWMutex
+	epLister listersv1.EndpointsLister
+	upToDate bool
+	store    *store
+}
+
+// NewKubeEndpointsFileConfigProvider returns a new KubeEndpointsFileConfigProvider
+func NewKubeEndpointsFileConfigProvider(config config.ConfigurationProviders) (ConfigProvider, error) {
```
Comment (clamoriniere) [inline pkg/autodiscovery/providers/kube_endpoints_file.go:64]: it seems that `config` is not used.

File: pkg/autodiscovery/listeners/kube_endpoints.go (hunk 4 above)
Comment (clamoriniere) [reply pkg/autodiscovery/listeners/kube_endpoints.go:69]: I just check and indeed it is an interface 👍

File: pkg/autodiscovery/providers/kube_endpoints_file.go (hunk 5 above)
Comment (ahmed-mez) [reply pkg/autodiscovery/providers/kube_endpoints_file.go]: removed the config argument. I can change the signature of the function in a separate PR.

**Review by clamoriniere — APPROVED** [review]

**Pull request merged**